### PR TITLE
(FACT-2921) Fix os.release fact on Linux

### PR DESCRIPTION
**Pull request by oanatmaria**
**Description of the problem:** Os.release fact is retrieved from the `/etc/os-release` file but Facter 3 reads other release files based on OS.
**Description of the fix** Retrieve os.release from the specific release file for every os.

### CODE REVIEW

File: lib/facter/resolvers/release_from_first_line.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+          output = Facter::Util::FileHelper.safe_read(release_file, nil)
+          return @fact_list[fact_name] = nil if output.nil?
+
+          @fact_list[:release] = retreive_version(output)
+
+          @fact_list[fact_name]
+        end
+
+        def retreive_version(output)
```
Comment (GabrielNagy) [inline lib/facter/resolvers/release_from_first_line.rb:31]: nit: retreive -> retrieve

File: lib/facter/resolvers/release_from_first_line.rb [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+          if output[/(Rawhide)$/]
+            'Rewhide'
```
Comment (GabrielNagy) [inline lib/facter/resolvers/release_from_first_line.rb:33]: nit: Rawhide?

**Review by GabrielNagy — COMMENTED** [review]
just some minor comments

File: lib/facter/util/facts/facts_utils.rb [hunk 3]
```diff
@@ -24,6 +24,30 @@ def discover_family(os)
...
+          versions = output.split('.')
+          fact_value = {}
+          fact_value['full'] = output
+          fact_value['major'] = versions[0]
+          fact_value['minor'] = versions[1] if versions[1]
+          fact_value
+        end
+
+        def release_hash_from_matchdata(data)
+          return unless data[1]
```
Comment (Filipovici-Andrei) [inline lib/facter/util/facts/facts_utils.rb:40]: you should also add a guard clause for when data is nil

File: lib/facter/facts/photon/os/release.rb [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+          [Facter::ResolvedFact.new(FACT_NAME, version),
+           Facter::ResolvedFact.new(ALIASES.first, version['major'], :legacy),
+           Facter::ResolvedFact.new(ALIASES.last, version['full'], :legacy)]
+        end
+
+        def from_specific_file
+          version = Facter::Resolvers::ReleaseWithRegex.resolve(:release,
+                                                                { release_file: '/etc/lsb-release',
+                                                                  regex: /DISTRIB_RELEASE="(\d+)\.(\d+)/ })
+          return if version[1].nil? || version[2].nil?
```
Comment (Filipovici-Andrei) [inline lib/facter/facts/photon/os/release.rb:24]: I think you also need a guard clause for when version is nil.

**Review by Filipovici-Andrei — CHANGES_REQUESTED** [review]

File: lib/facter/resolvers/aix/networking.rb [hunk 5]
```diff
@@ -9,7 +9,7 @@ class Networking < BaseResolver
         class << self
           private
 
-          def post_resolve(fact_name)
+          def post_resolve(fact_name, _options)
```
Comment (Filipovici-Andrei) [inline lib/facter/resolvers/aix/networking.rb:12]: why is this necessary?

File: lib/facter/framework/detector/os_detector.rb [hunk 6]
```diff
@@ -64,7 +64,14 @@ def detect_family
     Facter::Resolvers::OsRelease.resolve(:id_like)
   end
 
+  def detect_based_on_release_file
+    @identifier = :devuan if File.readable?('/etc/devuan_version')
+  end
+
   def detect_distro
+    detect_based_on_release_file
```
Comment (Filipovici-Andrei) [inline lib/facter/framework/detector/os_detector.rb:72]: I think this os is not as popular as the other ones, so I think that we should make this search probably before or after the SuseRelease. I noticed that we also extract the version, which is not used anywhere in the code (I could be wrong though).

**Review by Filipovici-Andrei — APPROVED** [review]

File: lib/facter/resolvers/release_with_regex.rb [hunk 7]
```diff
@@ -0,0 +1,34 @@
+# frozen_string_literal: true
+
+module Facter
+  module Resolvers
+    class ReleaseWithRegex < BaseResolver
```
Comment (gimmyxd) [inline lib/facter/resolvers/release_with_regex.rb:5]: This is the same as `SpecificReleseFile` + a regex. Since we exposed the `options`, why not have only a resolver which does the following:
 - reads version from a options[:release_file] or a default location of options[:release_file] is not sent
 - apply a regex if options[:regex] is sent

Maybe we can also compact `ReleaseFromFirstLine` into that.

File: lib/facter/facts/devuan/os/distro/release.rb [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+             Facter::ResolvedFact.new(ALIASES[2], release['minor'], :legacy)]
+          end
+
+          def construct_release(version)
+            versions = version.split('.')
+            release = {}
+            release['full'] = version
+            release['major'] = versions[0]
+            release['minor'] = versions[1] if versions[1]
+            release
```
Comment (gimmyxd) [inline lib/facter/facts/devuan/os/distro/release.rb:29]: You can use tap to avoid creating an empty has
```ruby
{}.tap do |release|
  release['full'] = version
  release['major'] = versions[0]
  release['minor'] = versions[1] if versions[1]
end
```

File: spec/fixtures/os_release_mint [hunk 9]
```diff
@@ -0,0 +1,2 @@
+DISTR_NAME=“Linuxmint”
```
Comment (GabrielNagy) [inline spec/fixtures/os_release_mint:1]: not sure if it matters but these are weird quotes `“` and `”`, should probably be `"`

File: lib/facter/facts/photon/os/release.rb [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+          major = version[1].to_s
+          minor = version[2].to_s
+          version_data = major + '.' + minor
+
+          fact_value = {}
+          fact_value['full'] = version_data
+          fact_value['major'] = major
+          fact_value['minor'] = minor
+          fact_value
```
Comment (gimmyxd) [inline lib/facter/facts/photon/os/release.rb:34]: Return the hash directly
```ruby
{
  'full' => version_data,
  'major' => major,
  'minor' => minor
}
```

File: lib/facter/util/facts/facts_utils.rb [hunk 11]
```diff
@@ -24,6 +24,23 @@ def discover_family(os)
...
+
+        def release_hash_from_string(output)
+          return unless output
+
+          versions = output.split('.')
+          fact_value = {}
+          fact_value['full'] = output
+          fact_value['major'] = versions[0]
+          fact_value['minor'] = versions[1] if versions[1]
+          fact_value
```
Comment (gimmyxd) [inline lib/facter/util/facts/facts_utils.rb:36]: You can use `tap` to make it shorter

**Review by GabrielNagy — APPROVED** [review]
looks good :+1:, just a minor fixture comment

File: lib/facter/resolvers/release_with_regex.rb (hunk 7 above)
Comment (oanatmaria) [reply lib/facter/resolvers/release_with_regex.rb]: `SpecificReleseFile` and `ReleaseWithRegex` are unified in `SpecificReleseFile`. The change is under the last commit for more visibility.

**Review by gimmyxd — APPROVED** [review]

**Comment by gimmyxd** [comment]
Looks good, @oanatmaria can you squash the commits before merging this?
